Honor the same set of proxy environment variables

ramboman opened the pull request:
# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->
Different parts of the project honor different sets of proxy environment variables. With this commit, all parts of the project will honor the same set of proxy environment variables.

# Context
<!-- Provide context. Reference open issues if available. -->
Related to: #10491

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Ericson2314 commented:
Thank you @ramboman! Can you move `proxyVariables` to a header in `libcmd` (and convert it to a `StringSet`) so `main.cc` and `nix-build.cc` can both share it?

ramboman commented:
Would you like `proxyVariables` to be in a separated header? If no, in which header do you want `proxyVariables` to go?

ramboman commented:
I am trying to create a `proxy.cc` and a `proxy.hh` in a separated project to
implement and test before moving them into `nix:/src/libcmd`.

Here is what I did for now:
- Create a separated project (directory) outside of `nix`
- Create `main.cc`:
```C++
#include <iostream>

#include "proxy.hh"

// print out a container
template<class Os, class K>
Os& operator<<(Os& os, const std::set<K>& v)
{
	os << '[' << v.size() << "] {";
	bool o{};
	for (const auto& e : v)
		os << (o ? ", " : (o = 1, " ")) << e;
	return os << " }";
}

int main(int argc, char * * argv)
{
	// src/nix-build/nix-build.cc
	nix::StringSet keepVars{
		"HOME", "XDG_RUNTIME_DIR", "USER", "LOGNAME", "DISPLAY",
		"WAYLAND_DISPLAY", "WAYLAND_SOCKET", "PATH", "TERM", "IN_NIX_SHELL",
		"NIX_SHELL_PRESERVE_PROMPT", "TZ", "PAGER", "NIX_BUILD_SHELL", "SHLVL"
	};
	keepVars.insert(nix::proxyVariables.begin(), nix::proxyVariables.end());
	std::cout << "keepVars: " << keepVars << std::endl;

	// src/nix/main.cc
	std::cout << "haveProxyConnection: " << (nix::haveProxyConnection() ? "true": "false") << std::endl;
	return 0;
}
```
- Create `proxy.hh`:
```C++
#pragma once
///@file

#include "types.hh"

namespace nix {

const StringSet lowercaseProxyVariables {
	"http_proxy", "https_proxy", "ftp_proxy", "all_proxy", "no_proxy"
};

/**
 * @return a StringSet with all proxy variables (uppercase and lowercase)
 */
StringSet getProxyVariables();

const StringSet proxyVariables = getProxyVariables();

/**
 * Heuristically check if there is a proxy connection by checking for defined
 * proxy variables.
 */
bool haveProxyConnection();

}

```
- Create `proxy.cc`:
```C++
#include "proxy.hh"

#include <algorithm>
#include <boost/algorithm/string.hpp>

#include "environment-variables.hh"

namespace nix {

StringSet getProxyVariables()
{
	StringSet proxyVariables = lowercaseProxyVariables;
	for (const auto & variable : lowercaseProxyVariables) {
		proxyVariables.insert(boost::to_upper_copy(variable));
	}

	return proxyVariables;
}

bool haveProxyConnection()
{
	const static StringSet excludeVariables {"no_proxy", "NO_PROXY"};
	StringSet variables;
	std::set_difference(
			proxyVariables.begin(), proxyVariables.end(),
			excludeVariables.begin(), excludeVariables.end(),
			std::inserter(variables, variables.begin()));
	for (const auto & variable : variables) {
		if (getEnv(variable).has_value()) {
			return true;
		}
	}
	return false;
}

}
```
- Create `types.hh`:
```C++
#pragma once

#include <set>
#include <string>

namespace nix {

typedef std::set<std::string> StringSet;

}
```
- Create `environment-variables.hh`:
```C++
#pragma once

#include <string>
#include <cstdlib>
#include <optional>

namespace nix {

std::optional<std::string> getEnv(const std::string & key)
{
	char * value = getenv(key.c_str());
	if (!value) return {};
	return std::string(value);
}

}
```
- Create `Makefile`:
```Makefile
CC_FILES = $(wildcard *.cc)
OBJS = $(CC_FILES:.cc=.o)
CC = g++
CCFLAGS = -Wall -g

all: main

main: $(OBJS)
	$(CC) $(CCFLAGS) -o $@ $^

%.o: %.cc
	$(CC) $(CCFLAGS) -c $< -o $@

clean:
	rm *.o
	rm main

.PHONY: all clean
```
- Execute this:
```
$ make
g++ -Wall -g -c main.cc -o main.o
g++ -Wall -g -c proxy.cc -o proxy.o
g++ -Wall -g -o main main.o proxy.o
$ ./main
Segmentation fault
```

When I merge `proxy.cc` in `proxy.hh`, it works but when they are separated,
there is a Segmentation fault I am unable to fix.

Some help is appreciated. Thank you!



N.B. Here is my system setup:
```
user@nix-impl:~$ uname --all
Linux nix-impl 6.6.25-1.qubes.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Apr  5 13:42:55 GMT 2024 x86_64 GNU/Linux
user@nix-impl:~$ 
user@nix-impl:~$ cat /etc/debian_version 
12.5
user@nix-impl:~$ 
user@nix-impl:~$ 
user@nix-impl:~$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/12/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 12.2.0-14' --with-bugurl=file:///usr/share/doc/gcc-12/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-12 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-12-bTRWOB/gcc-12-12.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-12-bTRWOB/gcc-12-12.2.0/debian/tmp-gcn/usr --enable-offload-defaulted --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.2.0 (Debian 12.2.0-14) 
user@nix-impl:~$ 
user@nix-impl:~$ 
user@nix-impl:~$ apt list --installed | grep build-essential

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

build-essential/stable,now 12.9 amd64 [installed]
user@nix-impl:~$ 
```

Ericson2314 commented:
I'm not really sure what happened, sorry, so I just fixed it myself and pushed a commit.

https://github.com/llvm/llvm-project/issues/61560 is the `clang-format` issue I am working around tw.

Ericson2314 commented:
Some of these variables are new, right? e.g. I don't see `ALL_PROXY` in any of the previous lists. Can we document where they come from?

ramboman commented:
Ericson2314:
> Some of these variables are new, right? e.g. I don't see `ALL_PROXY` in any of the previous lists. Can we document where they come from?

There were `all_proxy` variables in the lowercase proxy variables set:
```
$ cd <nix project directory>
$ git checkout 84e0c46
$ grep -r all_proxy
doc/manual/src/release-notes/rl-0.10.md:    `all_proxy` appropriately. Functions such as `fetchurl` in Nixpkgs
tests/nixos/nss-preload.nix:            "http_proxy" "https_proxy" "ftp_proxy" "all_proxy" "no_proxy"
src/libexpr/fetchurl.nix:    "http_proxy" "https_proxy" "ftp_proxy" "all_proxy" "no_proxy"
src/nix-build/nix-build.cc:        "http_proxy", "https_proxy", "ftp_proxy", "all_proxy", "no_proxy"
$
$ cd <nixpkgs project directory>
$ git checkout 8b4a7f28b
$ grep -r all_proxy
nixos/modules/config/networking.nix:          This option specifies the all_proxy environment variable.
nixos/modules/config/networking.nix:          all_proxy   = cfg.proxy.allProxy;
nixos/modules/services/networking/syncthing.nix:      all_proxy = mkOption {
nixos/modules/services/networking/syncthing.nix:          Overwrites the all_proxy environment variable for the Syncthing process to
nixos/modules/services/networking/syncthing.nix:          inherit (cfg) all_proxy;
lib/fetchers.nix:    "http_proxy" "https_proxy" "ftp_proxy" "all_proxy" "no_proxy"
```
I just added them where there were none, and added its uppercase counterpart.


Regarding the commit [535b7c7](https://github.com/NixOS/nix/pull/10611/commits/535b7c7018755c50f2de869cb762271696dff368):
- `src/nix-build/nix-build.cc` needs the `no_proxy` variable. `no_proxy` is one of the variables in the [documentation](https://github.com/NixOS/nix/blob/2f678331d59451dd6f1d9512cb6d92e4ecb9750f/doc/manual/src/installation/env-variables.md?plain=1#L54-L56).
- `src/nix/main.cc` does not use `no_proxy`, because the proxy variables here are used determine heuristically if a proxy connections are available. `no_proxy` does not serve that purpose.

ramboman commented:
All the other proxy variables in the `nix` project are listed [here](https://github.com/NixOS/nix/issues/10491#issuecomment-2068298239).

Ericson2314 commented on src/nix/network-proxy.hh:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ * Environment variables relating to network proxying. These are used by
+ * a few misc commands.
```
```suggestion
 * a few misc commands.
 *
 * See the "Environment section of https://curl.se/docs/manpage.html for details.
```

Ericson2314 commented on src/nix/network-proxy.hh:
```diff
@@ -0,0 +1,27 @@
+    "HTTPS_PROXY",
+    "FTP_PROXY",
+    "ALL_PROXY",
+    // clang-format on
```
What about `NO_PROXY` / `no_proxy`?

Ericson2314 commented on src/nix/network-proxy.hh:
```diff
@@ -0,0 +1,27 @@
+    "https_proxy",
+    "ftp_proxy",
+    "all_proxy",
+    "HTTP_PROXY",
```
https://curl.se/docs/manpage.html weird it says it is not a thing, but we can keep for back compat. Just add a comment maybe.

ramboman commented:
I was partially able to unify whatever that is related to the proxy environment variables under `src/libcmd/network-proxy.{cc,hh}` in [2d153b7](https://github.com/NixOS/nix/pull/10611/commits/2d153b7784c74bda83ffd7e8ab421f8800fdb681).

Here is what works:
- I can compile and install on my machine following these [instructions](https://nixos.org/manual/nix/stable/contributing/hacking#building-nix-with-flakes).
- I can heuristically check for proxy connections while avoiding to check the irrelevant `no_proxy` variables in that context:
```
$ pwd
/home/user/tmp/nix
$ which nix
/home/user/tmp/nix/outputs/out/bin/nix
$ cp -r /etc/nix outputs/out/etc
$ cat ../proxy.sh
set_vars () {
        export ALL_PROXY=127.0.0.1:8082
        export HTTP_PROXY=$ALL_PROXY
        export HTTPS_PROXY=$ALL_PROXY
        export FTP_PROXY=$ALL_PROXY
        export all_proxy=$ALL_PROXY
        export http_proxy=$all_proxy
        export https_proxy=$all_proxy
        export ftp_proxy=$all_proxy
}

unset_vars () {
        unset ALL_PROXY
        unset HTTP_PROXY
        unset HTTPS_PROXY
        unset FTP_PROXY
        unset all_proxy
        unset http_proxy
        unset https_proxy
        unset ftp_proxy
}

"$@"
$ . ../proxy.sh unset_vars
$ nix profile list
warning: you don't have Internet access; disabling some network-dependent features
<installed packages list>
$ no_proxy=https://www.google.com nix profile list
warning: you don't have Internet access; disabling some network-dependent features
<installed packages list>
$ all_proxy=127.0.0.1:8082 nix profile list
<installed packages list>
```

Here is what does not work:
- [CI / tests (ubuntu-latest) (pull_request)](https://github.com/NixOS/nix/actions/runs/8906699440/job/24459270709?pr=10611)
- [CI / tests (macos-latest) (pull_request)](https://github.com/NixOS/nix/actions/runs/8906699440/job/24459270837?pr=10611)
- Running `nix --experimental-features 'nix-command flakes' flake check -L` on my machine ends with a `137` exit code.

Feel free to:
- Revert back to [535b7c7](https://github.com/NixOS/nix/pull/10611/commits/535b7c7018755c50f2de869cb762271696dff368)
- Take inspiration from my [2d153b7](https://github.com/NixOS/nix/pull/10611/commits/2d153b7784c74bda83ffd7e8ab421f8800fdb681).
- etc.

ramboman commented:
I just fixed some clang-format non compliance. The CI / tests are fine now.

Ericson2314 commented on src/nix-build/nix-build.cc:
```diff
@@ -121,10 +121,9 @@ static void main_nix_build(int argc, char * * argv)
         "HOME", "XDG_RUNTIME_DIR", "USER", "LOGNAME", "DISPLAY",
         "WAYLAND_DISPLAY", "WAYLAND_SOCKET", "PATH", "TERM", "IN_NIX_SHELL",
-        "NIX_SHELL_PRESERVE_PROMPT", "TZ", "PAGER", "NIX_BUILD_SHELL", "SHLVL",
+        "NIX_SHELL_PRESERVE_PROMPT", "TZ", "PAGER", "NIX_BUILD_SHELL", "SHLVL"
```
trailing comma is actually good in my book, let's keep it.

Ericson2314 commented on src/nix/main.cc:
```diff
@@ -44,23 +45,6 @@ namespace nix {
 #ifdef _WIN32
 [[maybe_unused]]
 #endif
```
delete this too, it goes with the deleted function

Ericson2314 reviewed (changes requested):
Very close to being all good!

ramboman commented:
I just merged the `NixOS/master` branch into mine. The clang-format problem seems fixed at this point, in my case.

Ericson2314 commented:
Thanks @ramboman!

edolstra commented:
Note that this PR changes evaluation results. E.g. before:
```
$ nix eval --raw nixpkgs/ad7efee13e0d216bf29992311536fce1d3eefbef#hello.drvPath
/nix/store/img71klf2wzx867q5ldhf0zqc3wl0mxk-hello-2.12.1.drv
```
After:
```
$ nix eval --raw nixpkgs/ad7efee13e0d216bf29992311536fce1d3eefbef#hello.drvPath
/nix/store/7piwi9xdr81shbi2f9d3697f14ns2cf2-hello-2.12.1.drv
```
This is because it changes the `impureEnvVars` attribute passed to `builtin:fetchurl`.
